New freshness policies - docs for private preview (#29831)
dagster-io/dagster: docs/docs/guides/labs/freshness/index.md

policy = TimeWindowFreshnessPolicy.from_timedeltas(
    fail_window=timedelta(hours=24),
    warn_window=timedelta(hours=12)  # optional, must be less than fail_window
)

# Or, equivalently, from the base class
policy = InternalFreshnessPolicy.time_window(
    fail_window=timedelta(hours=24),
    warn_window=timedelta(hours=12)
)
```

The above policy states that there should be a successful materialization of the asset at least every 24 hours for it to be considered fresh. An asset that does not meet this condition will be considered failing its freshness policy.

There is an optional warning window (in this case, 12 hours). If the asset does not successfully materialize within this window, it will enter a `warning` freshness state.

## Setting Freshness Policies

### On Individual Assets

You can configure a freshness policy directly on an asset:

```python
from datetime import timedelta
from dagster._core.definitions.freshness import InternalFreshnessPolicy
from dagster._core.definitions.decorators.asset_decorator import asset
from dagster._core.definitions.asset_spec import AssetSpec

policy = InternalFreshnessPolicy.time_window(
    fail_window=timedelta(hours=24)
)

> @asset(internal_freshness_policy=policy)

Reviewer: You should call out and explain that this is called `internal_` only temporarily while in preview mode while we're moving towards removing support for legacy freshness policies.

Author: I added a whole section to explain legacy freshness policy and the "internal" prefix

---
title: 'Freshness Policies'
sidebar_position: 100
unlisted: True
---

> import Preview from '@site/docs/partials/\_Preview.md';

Reviewer 2: Instead of using the generic Preview partial, I recommend creating a freshness policy-specific partial, similar to the [dg & Components Preview partial](https://github.com/dagster-io/dagster/blob/master/docs/docs/partials/_DgComponentsPreview.md?plain=1), so you can add a more specific call to action -- I assume in this case, we want folks to reach out to their CSM if they have questions or feedback.

<Preview />

## Overview

Freshness policies help you understand which of your assets have materialized recently and which ones are running behind - a key component of asset health. Freshness policies also communicate expectations for data freshness, allowing downstream asset consumers to determine how often assets are expected to be updated.

For example, freshness policies can help identify stale assets caused by:

- Misconfigured `AutomationCondition`s
- Runs not being scheduled due to an upstream failure
- Runs taking longer than expected to complete

### Wait, isn't there already a `FreshnessPolicy`?

> Yes, there is an existing `FreshnessPolicy` API that has been deprecated for quite some time. We're opting to reuse the name for the new freshness APIs, which requires us to migrate off the legacy API.

Reviewer 2: ```suggestion
There is an existing `FreshnessPolicy` API that has been deprecated for quite some time. We're opting to reuse the name for the new freshness APIs, which requires us to migrate off the legacy API.
```

> To avoid naming conflicts with the legacy API during this migration, we've prefixed the new APIs, class names and method args with "internal" (for example, `InternalFreshnessPolicy` and the `internal_freshness_policy` arg on the `@asset` decorator). Once the migration is complete and the new APIs exit preview mode, the "internal" prefix will be removed and `FreshnessPolicy` will be the name of the new APIs.

Reviewer 2: ```suggestion
To avoid naming conflicts with the legacy API during this migration, we've prefixed the new APIs, class names, and method arguments with "internal", for example, `InternalFreshnessPolicy` and the `internal_freshness_policy` argument on the `@asset` decorator. Once the migration is complete and the new APIs exit preview mode, the "internal" prefix will be removed and `FreshnessPolicy` will be the name of the new APIs.
```

> ## Types of Freshness Policies

Reviewer 2: ```suggestion
## Freshness policy types
```

Currently, we support time window-based freshness policies, which are suitable for most use cases. We plan to add more policy types in the future.

> ### Time Window Policy

Reviewer 2: ```suggestion
### Time window policy
```

> A time window freshness policy is useful when you expect an asset to have new data or be recalculated with some frequency.

Reviewer 2: ```suggestion
A time window freshness policy is useful when you expect an asset to have new data, or be recalculated with some frequency. An asset that does not meet this condition will be considered failing its freshness policy. You can set an optional warning window on a freshness policy; if the asset does not successfully materialize within this window, it will enter a `warning` freshness state.

For example, the policy below states that there should be a successful materialization of the asset at least every 24 hours for it to be considered fresh, with a warning window of 12 hours:
```

<CodeExample path="docs_snippets/docs_snippets/guides/freshness/time_window_policy.py" language="python" />

> The above policy states that there should be a successful materialization of the asset at least every 24 hours for it to be considered fresh. An asset that does not meet this condition will be considered failing its freshness policy.

Reviewer 2: ```suggestion
```

> There is an optional warning window (in this case, 12 hours). If the asset does not successfully materialize within this window, it will enter a `warning` freshness state.

Reviewer 2: ```suggestion
```

> ## Setting Freshness Policies

Reviewer 2: ```suggestion
## Setting freshness policies
```

> ### On Individual Assets

Reviewer 2: ```suggestion
### On individual assets
```

You can configure a freshness policy directly on an asset:

<CodeExample path="docs_snippets/docs_snippets/guides/freshness/individual_asset_policy.py" language="python" />

> ### Across Multiple Assets

Reviewer 2: ```suggestion
### Across multiple assets
```

> To apply freshness policies to many or all assets in your deployment, you can use `map_asset_specs`:

Reviewer 2: ```suggestion
To apply freshness policies to multiple or all assets in your deployment, you can use `map_asset_specs`:
```

<CodeExample path="docs_snippets/docs_snippets/guides/freshness/multiple_assets_policy.py" language="python" />

You can also use `map_asset_specs` directly on the asset specs before creating a `Definitions` object:

<CodeExample path="docs_snippets/docs_snippets/guides/freshness/map_asset_specs_direct.py" language="python" />

:::caution

Applying a freshness policy in this way to an asset with an existing freshness policy (for example, if it was defined in the `@asset` decorator) will overwrite the existing policy.

:::

### Limitations

> Freshness policies are not currently supported for source observable assets (`SourceAssets`) and cacheable assets (`CacheableAssetsDefinition`).

Reviewer 2: ```suggestion
Freshness policies are not currently supported for source observable assets (<PyObject section="assets" module="dagster" object="SourceAsset" pluralize />) and cacheable assets (`CacheableAssetsDefinition`).
```

> ## Future Enhancements

Reviewer 2: ```suggestion
## Future enhancements
```

> - More types of freshness policies, including:

Reviewer 2: ```suggestion
- More freshness policy types, including:
```
